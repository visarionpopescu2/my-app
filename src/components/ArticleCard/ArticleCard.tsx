import {useContext, useState} from 'react';
import { Link} from 'react-router-dom';
import { Card, Popconfirm, Popover, Typography } from 'antd';
import { EditOutlined, ExportOutlined, DeleteOutlined } from '@ant-design/icons';
import { ARTICLE } from 'types';
import { ROUTES } from 'routes';
import { EditArticleModal } from 'components';
import { deleteArticle } from 'actions';
import NotificationContext from 'context/NotificationContext/NotificationContext';

type Props = {
  article: ARTICLE,
  onDeleteSuccess: (id: string) => void
  onEditSuccess: (article: ARTICLE) => void
}

function ArticleCard(props: Props) {
  const { article, onDeleteSuccess, onEditSuccess } = props;
  const [isLoading, setIsLoading] = useState(false);
  const [isFormOpen, setIsFormOpen] = useState(false);
  const { openNotification } = useContext(NotificationContext);
  
  const toggleEditForm = () => setIsFormOpen(!isFormOpen);
  
  const onDelete = async () => {
    setIsLoading(true);
    const response = await deleteArticle(article.id+'');
  
    if (response.error) {
      openNotification(response.error, 'error')
    } else {
      openNotification('The article has been successfully deleted', 'success');
      onDeleteSuccess(article.id+'');
    }
    setIsLoading(false);
  }
  
  return (
    <>
      <Card
        title={article.title}
        loading={isLoading}
        actions={[
          <EditOutlined key="edit" onClick={toggleEditForm} />,
          
          <Popconfirm
            key="delete"
            title='Delete Article'
            description='Please confirm you want to delete this article'
            onConfirm={onDelete}
          >
            <DeleteOutlined />
          </Popconfirm>,
          
          <Popover content="Go to article page">
            <Link to={`${ROUTES.Articles}/${article.id}`}>
              <ExportOutlined key="open" />,
            </Link>
          </Popover>
        ]}
      >
        <div style={{ textOverflow: 'ellipsis', overflow: 'hidden', whiteSpace: 'nowrap' }}>
          <Typography.Text>{article.body}</Typography.Text>
        </div>
      </Card>
      
      <EditArticleModal article={article} isOpen={isFormOpen} onClose={toggleEditForm} onSuccess={onEditSuccess} />
    </>
  )
}

export default ArticleCard;
