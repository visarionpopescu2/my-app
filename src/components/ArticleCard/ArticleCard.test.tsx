import { render } from '@testing-library/react';
import ArticleCard from './ArticleCard';
import {MemoryRouter} from 'react-router-dom';

const ARTICLE = { userId: 1, id: 1, title: 'title', body: 'body' };
const onDeleteSuccess = jest.fn();
const onEditSuccess = jest.fn();

describe('ArticleCard', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('should render ArticleCard component', () => {
    let view = render(
      <MemoryRouter>
        <ArticleCard article={ARTICLE} onDeleteSuccess={onDeleteSuccess} onEditSuccess={onEditSuccess} />
      </MemoryRouter>
    );
    expect(view.asFragment()).toMatchSnapshot();
  })
});
