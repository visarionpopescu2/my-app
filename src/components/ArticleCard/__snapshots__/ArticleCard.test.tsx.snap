// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleCard should render ArticleCard component 1`] = `
<DocumentFragment>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-2i2tap"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          title
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        style="text-overflow: ellipsis; overflow: hidden; white-space: nowrap;"
      >
        <span
          class="ant-typography css-dev-only-do-not-override-2i2tap"
        >
          body
        </span>
      </div>
    </div>
    <ul
      class="ant-card-actions"
    >
      <li
        style="width: 33.333333333333336%;"
      >
        <span>
          <span
            aria-label="edit"
            class="anticon anticon-edit"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="edit"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M257.7 752c2 0 4-.2 6-.5L431.9 722c2-.4 3.9-1.3 5.3-2.8l423.9-423.9a9.96 9.96 0 000-14.1L694.9 114.9c-1.9-1.9-4.4-2.9-7.1-2.9s-5.2 1-7.1 2.9L256.8 538.8c-1.5 1.5-2.4 3.3-2.8 5.3l-29.5 168.2a33.5 33.5 0 009.4 29.8c6.6 6.4 14.9 9.9 23.8 9.9zm67.4-174.4L687.8 215l73.3 73.3-362.7 362.6-88.9 15.7 15.6-89zM880 836H144c-17.7 0-32 14.3-32 32v36c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-36c0-17.7-14.3-32-32-32z"
              />
            </svg>
          </span>
        </span>
      </li>
      <li
        style="width: 33.333333333333336%;"
      >
        <span>
          <span
            aria-label="delete"
            class="anticon anticon-delete"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="delete"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
              />
            </svg>
          </span>
        </span>
      </li>
      <li
        style="width: 33.333333333333336%;"
      >
        <span>
          <a
            href="/articles/1"
          >
            <span
              aria-label="export"
              class="anticon anticon-export"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="export"
                fill="currentColor"
                fill-rule="evenodd"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M880 912H144c-17.7 0-32-14.3-32-32V144c0-17.7 14.3-32 32-32h360c4.4 0 8 3.6 8 8v56c0 4.4-3.6 8-8 8H184v656h656V520c0-4.4 3.6-8 8-8h56c4.4 0 8 3.6 8 8v360c0 17.7-14.3 32-32 32zM770.87 199.13l-52.2-52.2a8.01 8.01 0 014.7-13.6l179.4-21c5.1-.6 9.5 3.7 8.9 8.9l-21 179.4c-.8 6.6-8.9 9.4-13.6 4.7l-52.4-52.4-256.2 256.2a8.03 8.03 0 01-11.3 0l-42.4-42.4a8.03 8.03 0 010-11.3l256.1-256.3z"
                />
              </svg>
            </span>
            ,
          </a>
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
